### PR TITLE
global registration development

### PR DESCRIPTION
**Pull request by DamienGilliard**
New branch for registration development.

### CODE REVIEW

**Comment by 9and3** [comment]
Hello @DamienGilliard , just to update you that in the main now we have a logging system, you can rebase it here and starting implemented it in your functions i.e. for the registration. It's very basic but it will do for now, have a look at the [CONTRIBUTING file](https://github.com/diffCheckOrg/diffCheck/blob/main/CONTRIBUTING.md#logging). cheers!

image

**Comment by DamienGilliard** [comment]
@9and3 , @mariklad , @eleniv3d ,  this branch (which was renamed Global registration development) is open for review.

Latest major changes: 
* All methods of the globalRegistration class are now static
* All parameters for the registration, downsampling, ... are now passed as arguments of the globalRegistration methods, but with default values that I found worked well with our dataset.
* Thorough docstring documentation.

A new branch will be opened for refined registration methods such as ICP.

Thanks !

**Review by 9and3 — CHANGES_REQUESTED** [review]
Hello @DamienGilliard ! Thanks for the contribution! There is still quite some work imo and quite some fixes. Take your time to go through my comments and let me know if something is not clear.

File: src/diffCheck/registration/globalregistration.hh [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace diffCheck::registration
+{
+
+class GlobalRegistration
+{
+    public:
+    
+    static std::vector<double> ComputeP2PDistance(std::shared_ptr<geometry::DFPointCloud> source, 
+                                                  std::shared_ptr<geometry::DFPointCloud> target);
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/diffCheck/registration/globalregistration.hh:14]: Indeed, sorry.

File: src/diffCheck/registration/globalregistration.hh [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    @param source the source point cloud
+    @param target the target point cloud
+    @param voxelSize the size of the voxels used to downsample the point clouds
+    @param radiusKDTreeSearch the radius used to search for neighbors in the KDTree. It is used for the calculation of FPFHFeatures
+    @param maxNeighborKDTreeSearch the maximum number of neighbors to search for in the KDTree. It is used for the calculation of FPFHFeatures
+    @param maxCorrespondenceDistance the maximum distance between correspondences.
+    @param iterationNumber the number of iterations to run the RanSaC registration algorithm
+    @param maxTupleCount the maximum number of tuples to consider in the FPFH hyperspace
+    */
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/diffCheck/registration/globalregistration.hh:38]: got it

File: src/diffCheck/registration/globalregistration.cc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                                                                                                                     double voxelSize,
+                                                                                                                     double radiusKDTreeSearch,
+                                                                                                                     int maxNeighborKDTreeSearch,
+                                                                                                                     double maxCorrespondenceDistance)
+    {
+        std::shared_ptr<open3d::geometry::PointCloud> sourceO3D = source->Cvt2O3DPointCloud();
+        std::shared_ptr<open3d::geometry::PointCloud> targetO3D = target->Cvt2O3DPointCloud();
+
+        sourceO3D->VoxelDownSample(0.01);
+        targetO3D->VoxelDownSample(0.01);
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/diffCheck/registration/globalregistration.cc:130]: very sorry for that. Not pro on my part to let that one slip through

File: src/diffCheck/registration/globalregistration.cc [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                                                                                                                     int maxNeighborKDTreeSearch,
+                                                                                                                     double maxCorrespondenceDistance,
+                                                                                                                     int iterationNumber,
+                                                                                                                     int maxTupleCount)
+    {
+        std::shared_ptr<open3d::geometry::PointCloud> sourceO3D = source->Cvt2O3DPointCloud();
+        std::shared_ptr<open3d::geometry::PointCloud> targetO3D = target->Cvt2O3DPointCloud();
+
+        sourceO3D->VoxelDownSample(0.01);
+        targetO3D->VoxelDownSample(0.01);
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/diffCheck/registration/globalregistration.cc]: I am not sure I understand 100% : right now the function returns the transformation, not the transformed point cloud. I calculate the transform between the two downsampled point clouds, but I don't apply it, I return it as return value of the method. The transformation is applied in diffCheckApp.cc . Ok to return the transformation AND transformed point cloud as a tuple. Will do it tomorrow, 16/04.

File: src/diffCheck/registration/globalregistration.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/diffCheck/registration/globalregistration.hh]: I agree, I have to keep also in mind that it is documentation you see "per function" so if something is true across several functions, it must be repeated the same way across.

File: src/diffCheck/transformation/DFTransformation.cc [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#include "DFTransformation.hh"
+
+namespace diffCheck::transformation
+{
+
```
Comment (DamienGilliard) [inline src/diffCheck/transformation/DFTransformation.cc:5]: Would You say that it is the proper way to do this, or if the Class has no methods, we should not have any .cc file ?

**Comment by DamienGilliard** [comment]
@9and3  I believe all the issues noticed until now are fixed. The only issue I still have is to explain properly in the docstring the difference between the `O3DFastGlobalRegistrationFeatureMatching()`  and the  `O3DFastGlobalRegistrationBasedOnCorrespondence()` : The open3d documentation says that: 

_"In addition to FPFH-feature-based FGR, global registration can be performed with correspondence-based FGR via registration_fgr_based_on_correspondence. This method is useful if your correspondence frontend is different than FPFH, but you would still like to use FGR given a set of putative correspondences. It can be called with:
o3d.pipelines.registration.registration_fgr_based_on_correspondence(
        source_down, target_down, correspondence_set,
        o3d.pipelines.registration.FastGlobalRegistrationOption())"_ 

Only, in our case, I used a set of FPFH correspondences (the only ones easily computable with open3d), so in the end it is very similar to the method based on feature matching. Do we keep it for now knowing that there is currently little added value but can be expended later, or do we remove it, because what is not really useful right now should be discarded ?

**Comment by 9and3** [comment]
@DamienGilliard please have a look on the .gitignore if you are omitting all the vscode files, there is a conflict with `.vscode/settings.json`.

File: src/diffCheck/transformation/DFTransformation.cc (hunk 5 above)
Comment (9and3) [reply src/diffCheck/transformation/DFTransformation.cc:5]: Have a look at upcoming commit from my side for this point.

**Comment by 9and3** [comment]
@DamienGilliard , as discussed, let me know when you double checked the sanity of the  registration methods, thanks!

**Comment by DamienGilliard** [comment]
Changes made: 

- Bounding box member added to the DFPointCloud class
- Voxelize and radius search for global registrations now relative to the point cloud scale
- standard parameters of global registrations adapted

**Comment by DamienGilliard** [comment]
Sit Rep:
- The current code works well to register two half-overlapping point clouds, but still struggles with the dataset's point clouds
- long lines of code have been broken up to limit the width of the code: ```double absoluteVoxelSize =
 voxelSize * std::abs(sourceO3D->GetMaxBound().norm() - sourceO3D->GetMinBound().norm());```

**Comment by DamienGilliard** [comment]
@9and3  I believe the code is clean and ready to be reviewed, but as said in the slack, the registration is still not working perfectly with the dataset, but with a denser point cloud it works well ( 3 times successful with 75 000 points)

**Comment by 9and3** [comment]
Nice thanks for the corrections, lgtm now 👍 I am going to merge it...

**Comment by 9and3** [comment]
> (Sorry I couldn't resolve all the merge conflicts)

it's ok they were minor ones 👍

**Pull request merged**